### PR TITLE
Support Client Crendetials Grant

### LINKED ISSUE #15: Improve documentation on OAuth configuration
In the documentation under oauthConfig, the config values should be entered using strings, for example 'PASSWORD_CREDENTIALS_GRANT' for grant type. Expected from the lib are not strings but constants, like PASSWORD_CREDENTIALS_GRANT = 'password'. So in order to enter a working configuration, the constant have to be entered, or a 'password' string. There should be a description on how to import and use the constants.

### LINKED ISSUE #104: Support Client Grant Flow
See https://pages.github.bus.zalan.do/foundation/docs/oauth2-guide/oauth2-grants.html#client-grant and https://tools.ietf.org/html/rfc6749#section-4.4

The effort should be small since it is quite similar to the password grant. Furthermore, user should be encouraged to use the client grant instead of the password grant whenever possible (see linked resources above for explanation).

### PR DESCRIPTION
**Pull request by bzums**
Furthermore, tests are now using nock instead of explicitly starting a server.

close #104 
close #15

### CODE REVIEW

File: integration-test/oauth-tooling/getAccessToken.spec.ts [hunk 1]
```diff
@@ -98,216 +54,244 @@ describe('getAccessToken', () => {
...
-    });
+    const passwordCredentialsOAuthOptions = {
+      scopes: ['campaign.edit_all', 'campaign.read_all'],
+      accessTokenEndpoint: `${oAuthServerHost}${accessTokenEndpoint}`,
+      credentialsDir: 'integration-test/data/credentials',
+      grantType: PASSWORD_CREDENTIALS_GRANT
+    };
 
-    it('should be rejected if authorization header is invalid', function() {
+    it('should become the access token', () => {
```
Comment (Retro64) [inline integration-test/oauth-tooling/getAccessToken.spec.ts]: should return the access token if valid

File: integration-test/oauth-tooling/getAccessToken.spec.ts [hunk 2]
```diff
@@ -98,216 +54,244 @@ describe('getAccessToken', () => {
...
+      const promise = getAccessToken(passwordCredentialsOAuthOptions);
 
-      //then
-      return expect(promise).to.be.rejected;
+      // then
+      return expect(promise).to.become({ access_token: accessToken });
     });
 
-    it('should be rejected if credentials can not be read', function() {
+    it('should be rejected if credentials can not be read', () => {
```
Comment (Retro64) [inline integration-test/oauth-tooling/getAccessToken.spec.ts]: cannot

File: integration-test/oauth-tooling/getAccessToken.spec.ts [hunk 3]
```diff
@@ -98,216 +54,244 @@ describe('getAccessToken', () => {
...
-    });
+    const clientCredentialsOAuthOptions = {
+      scopes: ['campaign.edit_all', 'campaign.read_all'],
+      accessTokenEndpoint: `${oAuthServerHost}${accessTokenEndpoint}`,
+      credentialsDir: 'integration-test/data/credentials',
+      grantType: CLIENT_CREDENTIALS_GRANT
+    };
 
-    it('should return error message if authorization header is invalid', function() {
+    it('should become the access token', () => {
```
Comment (Retro64) [inline integration-test/oauth-tooling/getAccessToken.spec.ts]: should return the access token if valid

File: integration-test/oauth-tooling/getAccessToken.spec.ts [hunk 4]
```diff
@@ -98,216 +54,244 @@ describe('getAccessToken', () => {
...
+      const promise = getAccessToken(clientCredentialsOAuthOptions);
 
-      //then
-      return expect(promise).to.be.rejected;
+      // then
+      return expect(promise).to.become({ access_token: accessToken });
     });
 
-    it('should be rejected if credentials can not be read', function() {
+    it('should be rejected if credentials can not be read', () => {
```
Comment (Retro64) [inline integration-test/oauth-tooling/getAccessToken.spec.ts]: cannot

File: integration-test/oauth-tooling/getAccessToken.spec.ts [hunk 5]
```diff
@@ -98,216 +54,244 @@ describe('getAccessToken', () => {
...
+    const authCodeOAuthOptions = {
+      scopes: ['campaign.edit_all', 'campaign.read_all'],
+      accessTokenEndpoint: `${oAuthServerHost}${accessTokenEndpoint}`,
+      credentialsDir: 'integration-test/data/credentials',
+      grantType: AUTHORIZATION_CODE_GRANT,
+      code: validCode,
+      redirectUri: validRedirectUri
+    };
+
+    it('should become the access token', function() {
```
Comment (Retro64) [inline integration-test/oauth-tooling/getAccessToken.spec.ts]: should return the access token if valid

File: integration-test/oauth-tooling/getAccessToken.spec.ts [hunk 6]
```diff
@@ -98,216 +54,244 @@ describe('getAccessToken', () => {
...
+
+      // when
+      const promise = getAccessToken(authCodeOAuthOptions);
 
-    it('should be rejected, if grant type is unknown', function() {
+      //then
+      return expect(promise).to.become({access_token: '4b70510f-be1d-4f0f-b4cb-edbca2c79d41'});
+    });
+
+    it('should be rejected if credentials can not be read', () => {
```
Comment (Retro64) [inline integration-test/oauth-tooling/getAccessToken.spec.ts]: cannot

File: integration-test/oauth-tooling/getAccessToken.spec.ts [hunk 7]
```diff
@@ -98,216 +54,244 @@ describe('getAccessToken', () => {
...
+      accessTokenEndpoint: `${oAuthServerHost}${accessTokenEndpoint}`,
+      credentialsDir: 'integration-test/data/credentials',
+      grantType: REFRESH_TOKEN_GRANT,
+      refreshToken: validRefreshToken
+    };
 
-        return true;
-      })
-      .reply(HttpStatus.OK, responseObject);
+    it('should become the access token', () => {
```
Comment (Retro64) [inline integration-test/oauth-tooling/getAccessToken.spec.ts]: should return the access token if valid

**Comment by Retro64** [comment]
:+1:

**Comment by bzums** [comment]
👍

**Pull request merged**